conjure-undertow-processor/src/main/java/com/palantir/conjure/java/undertow/processor/generate/ConjureUndertowEndpointsGenerator.java

```diff
@@ -755,6 +757,26 @@ private static Optional<CodeBlock> getSafeLogging(
         }
     }
 
+    private static final ClassName LIST_NAME = ClassName.get(List.class);
+    private static final ClassName IMMUTABLE_LIST_NAME = ClassName.get(ImmutableList.class);
+    private static final ClassName SET_NAME = ClassName.get(Set.class);
+    private static final ClassName IMMUTABLE_SET_NAME = ClassName.get(ImmutableSet.class);
+
+    private static TypeName immutableCollection(TypeName input) {
+        if (input instanceof ParameterizedTypeName) {
+            ParameterizedTypeName parameterized = (ParameterizedTypeName) input;
+            if (LIST_NAME.equals(parameterized.rawType)) {
+                return ParameterizedTypeName.get(
+                        IMMUTABLE_LIST_NAME, parameterized.typeArguments.toArray(new TypeName[0]));
+            } else if (SET_NAME.equals(parameterized.rawType)) {
+                return ParameterizedTypeName.get(
+                        IMMUTABLE_SET_NAME, parameterized.typeArguments.toArray(new TypeName[0]));
+            }
+        }
+
+        return input;
+    }
```

Comment:
I could find a good place to put this shared logic, so I just copied it from `UndertowServiceHandlerGenerator` since the logic is fairly trivial and we may want these implementations to diverge in the future.

https://github.com/palantir/conjure-java/blob/260ed601c7faf034aa7cfba1dafc368078fa1795/conjure-java-core/src/main/java/com/palantir/conjure/java/services/UndertowServiceHandlerGenerator.java#L392-L412

I'm happy to move this to a shared helper method if there's somewhere we feel is appropriate.

Reply:
:+1: I don't mind code duplication, it's much nicer than unnecessarily coupling. The edge cases are dramatically different here vs pure conjure.